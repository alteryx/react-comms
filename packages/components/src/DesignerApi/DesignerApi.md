## Overview
The DesignerApi is a required component when you create your custom application with the UI-SDK, and design the app for Alteryx Designer. Every custom application must be wrapped in a DesignerApi. This lets the app access and update the relevant information in the parent application from anywhere within the component hierarchy.

When you implement the DesignerApi, it requires only one prop: `messages`. The messages prop is an object that includes any internationalized messages that you expect your app to access. The example below shows this in more detail. It also supports the `defaultConfig` prop, which is optional. It allows you to ensure a particular default state for your application before you've received your current tool configuration from Designer. Additionally, if you plan to use Eclipse Components, you must wrap your custom app in an AyxAppWrapper.

## Props Example
The DesignerApi expects you to provide your own translated messages for any languages you want to support, in the format shown below. It also supports a defaultConfig prop. This allows you to ensure a particular default state before you've received your current tool configuration from Designer.

``` jsx static
  import React from 'react'
  import { MessageAPI, DesignerApi } from '@ayx/react-comms'
  import { AyxAppWrapper } from '@ayx/eclipse-components'
  const messages = {
    en: {
      'example.label': 'Super awesome string.'
    },
    fr: {
      'example.label': 'Ficelle super génial.'
    },
    de: {
      'example.label': 'Super tolle Saite.'
    },
    es: {
      'example.label': 'Cadena super impresionante.'
    },
    pt: {
      'example.label': 'Corda super incrível.'
    },
    ja: {
      'example.label': '超素晴らしい文字列'
    },
    zh: {
      'example.label': '超棒的弦'
    }
  }

  <DesignerApi defaultConfig={{ Configuration: { count: 0 } }} messages={messages}>
    <AyxAppWrapper>
      Hello World
    </AyxAppWrapper>
  </DesignerApi>
```

## Update Your Data (Part 1)
The main roles of the DesignerApi are to expose model data and to update that model in both your custom app and the parent app. To do this, the DesignerAPI leverages React Context.

The DesignerApi gives any child component access to a `handleUpdateModel` method as well as the model itself. You can leverage this via the `useContext` hook. The example below uses the `handleUpdateModel` callback to update the model data for the custom app (in this case, a simple incrementer).

`handleUpdateModel` expects to be based on an object that contains one or all of the following keys: `Configuration`, `Secrets` and `Annotation`. Any other key will cause the update to fail and changes won't be persisted.

```js { "file": "../basicDemo.js" }
```

## Update Your Data (Part 2)
Most of the time, you're not going to be working with a simple incrementer, like the above example. In this example, we manage form data across many inputs in a nested model structure.

We've provided many different `onChange` callback examples to help you decide what you need to use. This will depend on the data structure and UI you've designed.  
```js { "file": "../advancedDemo.js" }
```

## Accessing Meta

The `DesignerApi` does some really nice cleanup for you when it comes to accessing your meta data. Meta will always be an array. Each index of the array represents a different list of fields from an input anchor. Accessing model.Meta[0] will give you the list of fields from the first input. If you wanted to access the 2nd field in that list, you'd chain an index of 1 to the fields property on the Meta. See below for an example:
```jsx static
const Child = () => {
  const [model, handleUpdateModel] = React.useContext(UiSdkContext);
  const greatMetaInfo = model.Meta[0].fields[1];
})
```

If you'd like to debug the meta data in more depth as it comes to you from Designer, simply add a `console.log(model.Meta)` from any component in your application where the model is accessible via context.

## Secrets

If your UI has password fields or other information that you'd consider sensitive, you can ensure it is obfuscated when stored in your XML by placing it in the `Secrets` key of the model. This key resolves to an object that you can place any key inside of. Each key should also be an object, with keys of `encryptionMode` and `text`. `text` references the secret value you'd like to have encoded and `encryptionMode` represents the obfuscation type you'd like to use: `obfuscation`, `machine`, or `user`. The default encryptionMode is `obfuscation` if it is not specified. When the tool configuration is saved off, anything you place into this Secrets key will automatically be obfuscated and unobfuscated upon the set and get configuration calls. See below: 

```jsx static
  const Child = () => {
    const [model, handleUpdateModel] = React.useContext(UiSdkContext);
    const { Secrets } = model;

    const handleChange = event => {
      handleUpdateModel({ Secrets: { password: { text: event.target.value, encryptionMode: 'obfuscation' }});
    }; 

    return (
      <input value={Secrets.password.text} onChange={handleChange}/>
    )
  }; 
```


