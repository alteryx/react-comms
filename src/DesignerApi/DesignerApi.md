## Overview
The DesignerApi is a required component when you create your custom application with the UI-SDK, and design the app for Alteryx Designer. Every custom application must be wrapped in a DesignerApi. This lets the app access and update the relevant information in the parent application from anywhere within the component hierarchy.

When you implement the DesignerApi, it expects to receive only one prop: `messages`. The messages prop is an object that includes any internationalized messages that you expect your app to have access to. The example below shows this in more detail. Additionally, if you plan to use UI-Core, you must wrap your custom app in an AyxAppWrapper.

## Props Example
The DesignerApi expects you to provide your own translated messages for any languages you'd like to support, in the format shown below.

``` jsx static
  import React from 'react'
  import { MessageAPI, Provider } from '@ayx/ui-sdk'
  import { AyxAppWrapper } from '@ayx/ui-core'
  const messages = {
    en: {
      'example.label': 'Super awesome string.'
    },
    fr: {
      'example.label': 'Ficelle super génial.'
    },
    de: {
      'example.label': 'Super tolle Saite.'
    },
    es: {
      'example.label': 'Cadena super impresionante.'
    },
    pt: {
      'example.label': 'Corda super incrível.'
    },
    ja: {
      'example.label': '超素晴らしい文字列'
    },
    zh: {
      'example.label': '超棒的弦'
    }
  }

  <DesignerApi messages={messages}>
    <AyxAppWrapper>
      Hello World
    </AyxAppWrapper>
  </Provider>
```

## Update Your Data (Part 1)
One of the main jobs of the DesignerApi is to expose model data and the ability to update that model in both your custom app and the parent app. To do this, the DesignerAPI leverages React Context.

The DesignerApi gives any child component access to a handleUpdateModel method as well as the model itself. You can leverage this through the useContext hook. The example below uses the handleUpdateModel callback to update the model data for the custom app (in this case, a simple incrementer). *Note:* You can only update the `Configuration` and `Annotation` keys of the model directly. Updates to any other key will fail. 

`handleModelUpdate` expects to be based on an object that contains one or both of these keys: Configuration and Annotation. Any other key will fail to cause an update.

```js { "file": "../basicDemo.js" }
```

## Update Your Data (Part 2)
Most of the time, you're not going to be working with a simple incrementer, like the above example. In this example, we manage form data across many inputs in a nested model structure.

We've provided many different onChange callback examples to help you decide what you might need to use. This will depend on the data structure and UI you've designed. 
```js { "file": "../advancedDemo.js" }

