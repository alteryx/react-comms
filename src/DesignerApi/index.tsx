/* eslint-disable react/require-default-props */
/* eslint-disable no-underscore-dangle */
/* eslint-disable react/forbid-prop-types */
import React, { useEffect, useState } from 'react';

import DesignerMessageApi, { IContext } from '../DesignerMessageApi';
import UiSdkContext, { IContextProviderProps } from '../Context';

interface IDesignerApiProps {
  messages: object;
  ctx?: any;
  children: React.ReactElement;
}

declare global {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  interface Window {
    Alteryx: IContext;
  }
}

let messageBroker;

const DesignerApi: React.FC = (props: IDesignerApiProps) => {
  const { messages = {} } = props;
  if (!messageBroker) {
    messageBroker = new DesignerMessageApi(props.ctx || window.Alteryx)
  }
  const [model, updateModel] = useState(messageBroker.model);
  const [appContext, updateAppContext] = useState(messageBroker.ayxAppContext);

  const handleUpdateModel = newModel => {
    // The reason all 3 of these are here is to work in all use cases for now, DesignerMessageAPI and DevHarness.
    // TODO: Refactor this to only be dependent on one call
    updateModel(newModel);
    messageBroker.model = newModel;
  };

  useEffect(() => {
    const receiveAppContext = data => {
      updateAppContext({ ...data });
    };
    const receiveModel = data => {
      updateModel({ ...data });
    };

    messageBroker.subscribe('MODEL_UPDATED', receiveModel);
    messageBroker.subscribe('APP_CONTEXT_UPDATED', receiveAppContext);
    return function cleanUp() {
      handleUpdateModel(messageBroker.model);
    };
  }, []);

  const contextProps: IContextProviderProps = {
    id: 'sdk-provider',
    value: [model, handleUpdateModel]
  };

  const { darkMode, locale, productTheme } = appContext || {};
  const appPropsToSpread = { messages, paletteType: darkMode ? 'dark' : 'light', theme: productTheme, locale };

  return (
    <UiSdkContext.Provider {...contextProps}>
      {React.cloneElement(props.children, { ...appPropsToSpread })}
    </UiSdkContext.Provider>
  );
};

export default DesignerApi;
