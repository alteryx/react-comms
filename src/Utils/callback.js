import { v4 as uuidv4 } from 'uuid';

const JsEvent = (context, Event, item = {}) => {
  const callbackRegistrationId = uuidv4();
  const callbackPromise = new Promise(res => {
    context.Gui.Callbacks[callbackRegistrationId] = (...theArgs) => {
      res.apply({}, theArgs);
    };
  });

  const actualObjectToSend = {
    Event,
    callback: `window.Alteryx.Gui.Callbacks['${callbackRegistrationId}']`,
    Callback: `window.Alteryx.Gui.Callbacks['${callbackRegistrationId}']`,
    ...item
  };

  context.JsEvent(JSON.stringify(actualObjectToSend));
  return callbackPromise;
};

export default JsEvent;
