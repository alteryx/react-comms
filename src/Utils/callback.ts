import { v4 as uuidv4 } from 'uuid';

import { IContext } from '../MessageApiBase/MessageApiBase';

export const JsEvent = (context: IContext, Event: string, item: object = {}): Promise<unknown> => {
  const callbackRegistrationId = uuidv4();
  const callbackPromise = new Promise(res => {
    context.Gui.Callbacks[callbackRegistrationId] = (...theArgs) => {
      res.apply({}, theArgs);
    };
  });

  const actualObjectToSend = {
    Event,
    callback: `window.Alteryx.Gui.Callbacks['${callbackRegistrationId}']`,
    Callback: `window.Alteryx.Gui.Callbacks['${callbackRegistrationId}']`,
    ...item
  };

  context.JsEvent(JSON.stringify(actualObjectToSend));
  return callbackPromise;
};
