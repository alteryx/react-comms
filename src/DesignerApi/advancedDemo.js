import React from 'react';
import {
  AyxAppWrapper,
  Button,
  Grid,
  Input,
  FormControl,
  FormHelperText,
  FormControlLabel,
  Checkbox,
  InputLabel
} from '@ayx/ui-core';

import UiSdkContext from '../Context/index.tsx';

import DesignerApi from './index.tsx';
import messages from './messages';

const initialModelState = {
  name: '',
  values: {
    field1: '',
    field2: '',
    checked: false
  }
};

const DesignerApiDemo = () => {
  const Child = () => {
    const [model, handleUpdateModel] = React.useContext(UiSdkContext);
    // Example of maintaing local state seperate from context state
    // Also allows us to "fake" having a model since there is no parent child app relationship here
    // In the real world, you'd likely only maintain state in one or the other.
    const [formData, updateForm] = React.useState(initialModelState);
    const resetModel = () => {
      updateForm(initialModelState);
      handleUpdateModel({ Configuration: { ...initialModelState } });
    };

    const submitModel = () => {
      alert(
        `This is the message that would be getting sent to your parent app: ${JSON.stringify({
          type: 'UPDATE_MODEL',
          payload: model
        })}`
      );
    };

    const handleNameChange = event => {
      const newModel = { ...formData, name: event.target.value };
      updateForm(newModel);
      handleUpdateModel({ Configuration: { ...newModel } });
    };

    const handleValueChange = (prop, event) => {
      const values = { ...formData.values, [prop]: event.target.value };
      const newModel = { ...formData, values };
      updateForm(newModel);
      handleUpdateModel({ Configuration: { ...newModel } });
    };

    const handleCheckedChange = () => {
      const values = { ...formData.values, checked: !formData.values.checked };
      const newModel = { ...formData, values };
      updateForm(newModel);
      handleUpdateModel({ Configuration: { ...newModel } });
    };

    return (
      <AyxAppWrapper>
        <Grid container spacing={3}>
          <Grid item md={3} sm={6} xs={12}>
            <FormControl>
              <InputLabel htmlFor="component-simple">Name</InputLabel>
              <Input id="component-simple" onChange={handleNameChange} value={formData.name} />
            </FormControl>
          </Grid>
          <Grid item md={3} sm={6} xs={12}>
            <FormControl aria-describedby="component-helper-text">
              <InputLabel htmlFor="component-helper">Value 1</InputLabel>
              <Input
                id="component-helper"
                onChange={e => handleValueChange('field1', e)}
                value={formData.values.field1}
              />
              <FormHelperText id="component-helper-text">Some important helper text</FormHelperText>
            </FormControl>
          </Grid>
          <Grid item md={3} sm={6} xs={12}>
            <FormControl>
              <InputLabel htmlFor="component">Value 2</InputLabel>
              <Input id="component" onChange={e => handleValueChange('field2', e)} value={formData.values.field2} />
              <FormHelperText>Some more important helper text</FormHelperText>
            </FormControl>
          </Grid>
          <Grid item>
            <FormControl component="fieldset">
              <FormControlLabel
                checked={formData.values.checked}
                control={<Checkbox />}
                label="Checked"
                labelPlacement="top"
                onChange={handleCheckedChange}
                value="checked"
              />
            </FormControl>
          </Grid>
          <Grid item xs={3}>
            <Button color="primary" onClick={submitModel} variant="contained">
              Submit Model
            </Button>
          </Grid>
          <Grid item xs={3}>
            <Button onClick={resetModel} variant="contained">
              Reset Model
            </Button>
          </Grid>
        </Grid>
      </AyxAppWrapper>
    );
  };

  return (
    <DesignerApi defaultConfig={{ Configuration: { ...initialModelState } }} messages={messages}>
      <Child />
    </DesignerApi>
  );
};

<DesignerApiDemo />;
