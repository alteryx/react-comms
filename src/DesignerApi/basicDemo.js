import React from 'react';
import { AyxAppWrapper, Button, Grid, Typography } from '@ayx/ui-core';

import UiSdkContext from '../Context/index.tsx';

import DesignerApi from './index.tsx';
import messages from './messages';

const DesignerApiDemo = () => {
  const Child = () => {
    const [model, handleUpdateModel] = React.useContext(UiSdkContext);
    const incrementCount = () => {
      const newModel = { ...model };
      newModel.count++;
      handleUpdateModel(newModel);
    };
    return (
      <AyxAppWrapper>
        <Grid container>
          <Grid item>
            <Button id="child" onClick={incrementCount} variant="contained">
              <Typography>Click me to increment the count</Typography>
            </Button>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h1">{model.count}</Typography>
          </Grid>
        </Grid>
      </AyxAppWrapper>
    );
  };

  return (
    <DesignerApi messages={messages}>
      <Child />
    </DesignerApi>
  );
};

<DesignerApiDemo />;
