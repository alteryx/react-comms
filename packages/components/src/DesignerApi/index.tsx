/* eslint-disable react/require-default-props */
/* eslint-disable no-underscore-dangle */
/* eslint-disable no-console */
/* eslint-disable react/forbid-prop-types */
import React, { useEffect, useState, useCallback } from 'react';
import merge from 'deepmerge';
import deepEqual from 'deep-equal';

import DesignerMessageApi from '../DesignerMessageApi';
import MicroAppMessageApi from '../MicroAppMessageApi';
import { IContext, IModel } from '../Utils/types';
import { SUBSCRIPTION_EVENTS } from '../Utils/constants';
import UiSdkContext, { IContextProviderProps } from '../Context';

interface IDesignerApiProps {
  messages: object;
  ctx?: IContext;
  defaultConfig?: object;
  children: React.ReactElement;
}

declare global {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  interface Window {
    Alteryx: IContext;
  }
}

let messageBroker: DesignerMessageApi | MicroAppMessageApi;

const DesignerApi: React.FC<IDesignerApiProps> = (props: IDesignerApiProps) => {
  const { messages = {}, defaultConfig = {} } = props;
  if (!messageBroker) {
    messageBroker =
      window.Alteryx && window.Alteryx.AlteryxLanguageCode
        ? new DesignerMessageApi(props.ctx || window.Alteryx)
        : new MicroAppMessageApi();
  }
  const mergedState: IModel = merge(messageBroker.model, defaultConfig);
  const [model, updateModel] = useState(mergedState);
  const [appContext, updateAppContext] = useState(messageBroker.ayxAppContext);

  useEffect(() => {
    const modelChanged = !deepEqual(messageBroker.model, model, { strict: true });
    if (modelChanged) {
      // Just update the messageBroker model whenever state updates.
      messageBroker.model = model;
      messageBroker instanceof MicroAppMessageApi
        ? messageBroker.sendMessage(SUBSCRIPTION_EVENTS.MODEL_UPDATED, model)
        : window.Alteryx.model = model;
    }
  }, [model])

  useEffect(() => {
    const receiveAppContext = data => {
      updateAppContext({ ...data });
    };
    const receiveModel = data => {
      updateModel(merge(model, data));
    };

    messageBroker.subscribe(SUBSCRIPTION_EVENTS.MODEL_UPDATED, receiveModel);
    messageBroker.subscribe(SUBSCRIPTION_EVENTS.AYX_APP_CONTEXT_UPDATED, receiveAppContext);
    return function cleanUp() {
      updateModel(messageBroker.model);
    };
  }, []);

  const getContextValue = useCallback(() => [model, updateModel], [model, updateModel]);
  const contextProps: IContextProviderProps = {
    id: 'sdk-provider',
    value: getContextValue()
  };
  
  const { darkMode, locale, productTheme } = appContext || {};
  const appPropsToSpread = { messages, paletteType: darkMode ? 'dark' : 'light', theme: productTheme, locale };

  return (
    <UiSdkContext.Provider {...contextProps}>
      {React.cloneElement(props.children, { ...appPropsToSpread })}
    </UiSdkContext.Provider>
  );
};

export default DesignerApi;
