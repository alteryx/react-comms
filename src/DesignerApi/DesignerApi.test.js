/* eslint-disable no-underscore-dangle */
import React from 'react';
import { shallow, mount } from 'enzyme';

import UiSdkContext from '../Context/index.tsx';
import * as callback from '../Utils/callback.ts';

import DesignerApi from './index.tsx';

describe('DesignerApi', () => {
  beforeAll(() => {
    window.Alteryx = {
      AlteryxLanguageCode: 'en',
      Gui: {
        SetConfiguration: jest.fn(),
        GetConfiguration: jest.fn()
      },
      JsEvent: jest.fn()
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
  it('should be able to render on the DOM', () => {
    const Child = () => {
      const [model, handleUpdateModel] = React.useContext(UiSdkContext);
      if (model.Annotation !== 'foo') handleUpdateModel({ Annotation: 'foo' });
      return <div id="child">{model.Annotation}</div>;
    };
    const wrapper = shallow(
      <DesignerApi ctx={window.Alteryx}>
        <Child />
      </DesignerApi>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should be able to render on the DOM with a defaultConfig', async () => {
    const Child = () => {
      const [model] = React.useContext(UiSdkContext);
      return <div id="child">{model.Annotation}</div>;
    };
    const wrapper = await shallow(
      <DesignerApi ctx={window.Alteryx} defaultConfig={{ Annotation: 'foo' }}>
        <Child />
      </DesignerApi>
    );

    const valueProp = wrapper.find('#sdk-provider').prop('value');
    expect(valueProp[0]).toEqual({
      ToolName: '',
      ToolId: undefined,
      Annotation: 'foo',
      Configuration: {},
      Secrets: {},
      Meta: [],
      srcData: {}
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render AyxAppWrapper with appropriate props if there is model data', () => {
    const Child = () => {
      const [model, handleUpdateModel] = React.useContext(UiSdkContext);
      if (model.Annotation !== 'foo') handleUpdateModel({ Annotation: 'foo' });
      return <div id="child">{model.Annotation}</div>;
    };
    const wrapper = shallow(
      <DesignerApi ctx={window.Alteryx}>
        <Child />
      </DesignerApi>
    );
    expect(wrapper.find('#sdk-provider').children().props()).toEqual({
      paletteType: 'light',
      theme: {},
      locale: 'en',
      children: undefined,
      messages: {}
    });
  });

  it('should render the sdk-provider with a model when an event is emitted', () => {
    const Child = () => {
      const [model] = React.useContext(UiSdkContext);
      return <div id="child">{model.Annotation}</div>;
    };
    const wrapper = shallow(
      <DesignerApi ctx={window.Alteryx}>
        <Child />
      </DesignerApi>
    );
    const valueProp = wrapper.find('#sdk-provider').prop('value');
    expect(valueProp).toHaveLength(2);
    expect(valueProp[0]).toEqual({
      ToolName: '',
      ToolId: undefined,
      Annotation: '',
      Configuration: {},
      Secrets: {},
      Meta: [],
      srcData: {}
    });
    expect(valueProp[1]).toBeInstanceOf(Function);
  });

  it('should use the context hook to update the providers model', () => {
    const Child = () => {
      const [model, handleUpdateModel] = React.useContext(UiSdkContext);
      if (model.Annotation !== 'foo') {
        handleUpdateModel({ Annotation: 'foo' });
      }
      return <div id="child">{model.Annotation}</div>;
    };
    const wrapper = mount(
      <DesignerApi ctx={window.Alteryx}>
        <Child />
      </DesignerApi>
    );

    expect(wrapper.find('#child').text()).toEqual('foo');
  });

  it('should use the current model state as the payload when GetConfiguration is called', async () => {
    const spyJsEvent = jest.spyOn(callback, 'JsEvent');
    const expected = {
      Configuration: {
        Configuration: {
          Secrets: {}
        },
        Annotation: 'foo'
      }
    };

    const Child = () => {
      const [model, handleUpdateModel] = React.useContext(UiSdkContext);
      if (model && model.Annotation !== 'foo') {
        handleUpdateModel({ Annotation: 'foo' });
      }
      return <div id="child">{model.Annotation}</div>;
    };
    const wrapper = shallow(
      <DesignerApi ctx={window.Alteryx}>
        <Child />
      </DesignerApi>
    );
    wrapper.update();
    await window.Alteryx.Gui.GetConfiguration();
    expect(spyJsEvent).toHaveBeenCalledWith(window.Alteryx, 'GetConfiguration', expected);
  });

  it('should fail to update if handleUpdateModel is passed a bad key', () => {
    let rendered = false;
    const Child = () => {
      const [model, handleUpdateModel] = React.useContext(UiSdkContext);
      if (!rendered) {
        handleUpdateModel({ Annotation: 'not-foo', badStuff: 'okay' });
        rendered = true;
      }
      return <div id="child">{model.Annotation}</div>;
    };
    const wrapper = mount(
      <DesignerApi ctx={window.Alteryx}>
        <Child />
      </DesignerApi>
    );

    expect(wrapper.find('#child').text()).toEqual('foo');
  });

  it('should fail to update the secrets key if it is passed an invalid secret', () => {
    const Child = () => {
      const [model, handleUpdateModel] = React.useContext(UiSdkContext);
      if (!model.Secrets.shouldSave) {
        handleUpdateModel({ Secrets: { password: { noGood: 'no update' }, shouldSave: 'okay' } });
      }
      return <div id="child">{model.Secrets.shouldSave}</div>;
    };

    const wrapper = mount(
      <DesignerApi ctx={window.Alteryx}>
        <Child />
      </DesignerApi>
    );
    expect(wrapper.find('#child').text()).toEqual('okay');
  });
});
