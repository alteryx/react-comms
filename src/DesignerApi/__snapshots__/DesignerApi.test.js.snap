// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DesignerApi should be able to render on the DOM 1`] = `
<ContextProvider
  id="sdk-provider"
  value={
    Array [
      Object {
        "Annotation": "",
        "Configuration": Object {},
        "Meta": Array [],
        "Secrets": Object {},
        "ToolId": undefined,
        "ToolName": "",
        "srcData": Object {},
      },
      [Function],
    ]
  }
>
  <Child
    locale="en"
    messages={Object {}}
    paletteType="light"
    theme={Object {}}
  />
</ContextProvider>
`;

exports[`DesignerApi should be able to render on the DOM with a defaultConfig 1`] = `
<ContextProvider
  id="sdk-provider"
  value={
    Array [
      Object {
        "Annotation": "foo",
        "Configuration": Object {},
        "Meta": Array [],
        "Secrets": Object {},
        "ToolId": undefined,
        "ToolName": "",
        "srcData": Object {},
      },
      [Function],
    ]
  }
>
  <Child
    locale="en"
    messages={Object {}}
    paletteType="light"
    theme={Object {}}
  />
</ContextProvider>
`;
